userns: add support for auto

giuseppe opened the pull request:
automatically pick an empty range and create an user namespace for the container.
    
For root containers, it is necessary to specify an entry in the /etc/subuid and /etc/subgid files to use for picking the range of available IDs.  This is necessary to avoid collisions with IDs used for rootless containers.  This setting is ignored for rootless containers, since it is not possible to use arbitrary IDs, and the initial set is always picked by the IDs assigned to the rootless user.
    
When using auto userns, a container will use a range of IDs that is not used by any other container user namespace, also those that are not using auto userns, this is checked at creation time.

A successive container that doesn't use auto userns feature can still collide with IDs used by an auto userns container.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rhatdan commented on storage.conf:
```diff
@@ -46,6 +46,14 @@ additionalimagestores = [
+# configured to automatically create an user namespace can still overlap with containers
+# having an explicit mapping set.
+# This setting is ignored when running as rootless.
+# root-auto-userns-user = "storage"
```
I would prefer default should be "containers",

vrothberg replied:
:+1: on "containers"

rhatdan commented on userns.go:
```diff
@@ -0,0 +1,220 @@
+		}
+	} else {
+		if username == "" {
+			username = "storage"
```
Containers, and should be a constant.

giuseppe replied:
it is the name we are already using for `RemapUser`.

I am fine with changing it but I think we should be consistent.  Should we change the default for `RemapUser` too?

vrothberg replied:
SGTM

rhatdan commented on userns.go:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	if size <= 0 {
+		size = 4096
```
Should this be defined in containers/storage?

giuseppe replied:
should we allow to override it (via storage.conf) as part of this PR or can it be done later?

giuseppe replied:
I've added  a patch that allows setting the minimum and maximum value for the user namespace.

The maximum size is necessary as a malicious image could consume the entire space.

rhatdan commented:
@vrothberg @saschagrunert @nalind @TomSweeneyRedHat PTAL

rhatdan commented:
I think we need a better description of what we are after here.

The basic idea is to allow tools like Podman and CRI-O to automatically grab a unique UID/GID range for a user namespace large enough for the container, based on the UIDs/GIDs defined in the image.  We need to go through all existing containers and make sure of no overlap.

Also we will preallocate a range of uids in the /etc/subuid and /etc/subgid file when tools like CRI-O, Buildah, Podman get installed.

Something like:
```
cat /etc/subuid
3267:100000:8000
containers:2147483646:2147483648
```

Then the container engines can grab ranges from 2147483646 on, and useradd and systemd can continue to use the bottom range of UIDs.

giuseppe commented:
> I think we should also check the highest uid in the /etc/passwd defined in the container as well, to make sure that is not greater then the range of UIDs.

There are other heuristic we should add (like check `/etc/sub*id`) but we need to do it at pull time when we first create the layer.

giuseppe commented:
libpod implementation: https://github.com/containers/libpod/pull/5478

rhatdan commented:
I am fine with doing it later.

I think we need more tests.  How do we check what happens when we run out of ranges?

giuseppe commented:
> I think we need more tests. How do we check what happens when we run out of ranges?

I'll add more tests

vrothberg commented on docs/containers-storage.conf.5.md:
```diff
@@ -76,6 +76,15 @@ The `storage.options` table supports the following options:
+  Auto-userns-min-size is the minimum size for an user namespace created automatically.
+
+**auto-userns-max-size**=65536
+  Auto-userns-max-size is the maximum size for an user namespace created automatically.
```
Here and above s/an user/a user/

vrothberg replied:
And some of the code comments.

vrothberg commented on store.go:
```diff
@@ -1151,6 +1179,34 @@ func (s *store) CreateContainer(id string, names []string, image, layer, metadat
+		if err != nil {
+			return nil, err
+		}
+	}
```
Could we turn this code into a anonymous function/lambda? This would reduce the critical section and hence the time we keep the lock (and ultimately speed it up a bit more).

vrothberg replied:
```golang
uidMap, gidMap, err := func () ... {
   lock()
   defer unlock()
}

giuseppe replied:
we need to serialize the entire function.  We can release the lock only after the container is created in the store to prevent another process to pick the same range.

vrothberg commented on store.go:
```diff
@@ -3305,6 +3361,9 @@ func copyStringInterfaceMap(m map[string]interface{}) map[string]interface{} {
 // defaultConfigFile path to the system wide storage.conf file
 const defaultConfigFile = "/etc/containers/storage.conf"
 
+const AutoUserNsMinSize = 1024
```
The docs state `4096`. Which value is the right one?

giuseppe replied:
updated the docs to the correct value

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+	"github.com/sirupsen/logrus"
+)
+
+func getAutoUserNsMappings(username string) ([]idtools.IDMap, []idtools.IDMap, error) {
```
A comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+	return uids, gids, nil
+}
+
+func (s *store) getAvailableMappings() ([]idtools.IDMap, []idtools.IDMap, error) {
```
A comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+		usedGIDs = append(usedGIDs, c.GIDMap...)
+	}
+
+	if size <= 0 {
```
A (verbose) comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+	return findAvailableRange(size, size, availableUIDs, availableGIDs, usedUIDs, usedGIDs)
+}
+
+func findAvailableRange(sizeUID, sizeGID uint32, availableUIDs, availableGIDs, usedUIDs, usedGIDs []idtools.IDMap) ([]idtools.IDMap, []idtools.IDMap, error) {
```
A comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+	return UIDMap, GIDMap, nil
+}
+
+func findAvailableIDRange(size uint32, availableIDs, usedIDs []idtools.IDMap) ([]idtools.IDMap, error) {
```
A comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+		}
+		avail = append(avail, n)
+	}
+	for _, u := range usedIDs {
```
A (verbose) comment would be great.

vrothberg commented on userns.go:
```diff
@@ -0,0 +1,235 @@
+	return nil, errors.New("could not find enough available IDs")
+}
+
+func subtractIDs(avail idtools.IDMap, used idtools.IDMap) []idtools.IDMap {
```
A comment would be great.

giuseppe commented:
thanks for the review.  I've pushed another version that addresses the points above.

I've added a patch to change the default user from `storage` to `containers`

giuseppe commented:
marked WIP until Podman tests pass

vrothberg commented:
@giuseppe, mind squashing the commits into one?

giuseppe commented:
> @giuseppe, mind squashing the commits into one?

isn't clearer to keep the changes separate?

The first patch adds the basic support for auto-userns, the second one extends auto-userns with a new option (there are likely more patches to be added in the future for other options) and the the third one changes an existing setting.  The current third patch should probably be moved first before the other two patches

giuseppe commented:
> @giuseppe, mind squashing the commits into one?

I've squashed the two patches for auto-userns and left the one that changes `storage` -> `containers`

giuseppe commented:
@rhatdan ready for review

rhatdan commented:
LGTM
